DEVELOPMENT_BEST_PRACTICE.md

```diff
@@ -0,0 +1,308 @@
+
+### Concurrency
+
+Concurrency introduces harder to rationalize execution relative to serialized programs and is easy to get wrong. Concurrency doesn't necessarily speed up a program either. With these points in mind, prefer synchronous APIs instead leaving concurrency to the caller. Not only does this eliminate a whole host of problems, it ensures your APIs are predictable and easier to use.
+
+Dave Cheney has made some excellent [comments](https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#concurrency) on concurrency in Go.
+
+#### Channels
+
+If you find the need for channels prefer accepting them as parameters. Accepting channels as a parameter allows consumers to specify channel properties, such as buffered or unbuffered, creating a more flexible API. 
+
+Clearly document channel ownership thereby documenting who is responsible for closing it. If there are multiple writers, no-one can close it therefore no-one owns it. Note channels don't _need_ to be closed, they will be garbage collected once the necessary scoping conditions are met.
+
+Clearly document channel behavior. For example, what does your algorithm do if it can't send data on a channel? Will it drop data, block until the channel is writable, or something else?
+
+Avoid using channels for data mutation operations where a traditional mutex would suffice. 
+
+###  Comments
+
+Functions and variables should be named so that their purpose is clear. However, good naming and well structured code isn't justification to not document APIs. All public package APIs should be documented including those that may seem trivial. This, at minimum, helps with code documentation generation. Additionally, each package should have a clear purpose that can be documented with examples using the [package documentation feature of Go](https://go.dev/doc/effective_go#commentary).
+
+Through good documentation that we reflect on and update regulalry, we can clearly communicate with maintainers what the original intent is making decisions easier when new capabilities are required.
+
+### Returns
+
+Return early instead of nesting deeply. The `else` statement is rarely required when following this rule. For example:
+
+```go
+func foo(bar string) error {
+    if bar == "" {
+        return errors.New("bar is empty")
+    }
+
+    // do something
+
+    return nil
+}
+```
+
+In returning early, we also create clearer code that has lower [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) that can be the cause for bugs.
+
+Never use naked returns. Go allows return types to be specified with an identifier. The identifier is a scoped variable in the context of the function and its value can be set during program execution. Doing so, however, often creates unnecessraily difficult to rationalize logic.
+
+Avoid named returns. If you don't specify names for your return values you can't accidentally use naked returns. 
```

Comment:
I disagree with this. I think named returns are very useful when returning things like `(string, string)` and I think we should encourage them.

I would agree with avoiding them when unnecessary (eg. `ReadFileName() (string, error)`). Also 100% agree with "Never use naked returns".

Reply:
Thats why this is `avoid`. Common sense needs to be used and if you have a function that returns the same type twice then clearly named returns help clarify whats what. Thats pretty much the only case I thought about that warrants named returns and caused my to choose the wording 'avoid'.

Reply:
What would you think about replacing "avoid" with something like:
```suggestion
Use named returns only if strictly necessary for disambiguation (for example, a function that returns the same type twice). If you don't specify names for your return values you can't accidentally use naked returns. 
```